File: tags/or_0_1_0-m3/openregistry-webapp/src/etc/OpenRegistryActivationRESTClient.py
#
# Copyright (C) 2009 Jasig, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# httplib2 library - needs to be installed: http://code.google.com/p/httplib2/
import httplib2

def verifyAndInvalidateActivationKey(activation_key_resource_client, activation_key_resource_uri):
	print 'Processing the activation key..............\n'
	
	#Verify - HTTP GET
	resp, response_content = activation_key_resource_client.request(activation_key_resource_uri, "GET")
	if resp['status'] == '204':
		print 'The activation is valid'
	else:
		print response_content
	
	#Now invalidate - HTTP DELETE
	response_tuple = activation_key_resource_client.request(activation_key_resource_uri, "DELETE")
	print "The invalidation of the key returned HTTP status code: %s" % response_tuple[0]['status']		

#Main script body
open_registry_activation_api = httplib2.Http()
open_registry_activation_api.add_credentials('rats', 'RAT1234')

#Perform an HTTP POST request
response_tuple = open_registry_activation_api.request("https://dev-registry.rutgers.edu/api/v1/people/NETID/HR01/activation", "POST")
print "Generate Activation Key call status: %s" % response_tuple[0]['status']
print "New Activation Key resource Location: %s" % response_tuple[0]['location']
print "New Activation Key value: %s\n" % response_tuple[0]['location'].rpartition('/')[2]

#Further process the activation key
verifyAndInvalidateActivationKey(open_registry_activation_api, response_tuple[0]['location'])

File: trunk/openregistry-webapp-jasig/src/etc/OpenRegistryActivationRESTClient.m
#include <CoreFoundation/CoreFoundation.h>
#include "ASIHTTPRequest.h"

/* Objective-C sample HTTP cleint for Open Registry Activation RESTful resource.
 * Uses 'ASIHTTPRequest library. for installation instructions see: http://allseeing-i.com/ASIHTTPRequest/
 */

//Function prototypes
static int verify_and_invalidate_activation_key(NSURL *);
static ASIHTTPRequest * setup_request(NSURL *);

int main (int argc, const char * argv[]) {
    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
    
	ASIHTTPRequest *request = setup_request([NSURL URLWithString:@"https://dev-registry.rutgers.edu/api/v1/people/NETID/HR01/activation"]);

    //Perform an HTTP POST request
    [request setRequestMethod:@"POST"];
    [request start];
    
    NSError *error = [request error];
    if(error) {
        NSLog([error description]);
        [pool drain];
        return 1;
    }
    
    printf("Generate Activation Key call status: %d\n", [request responseStatusCode]);
    printf("New Activation Key resource Location: %s\n", [[[request responseHeaders] objectForKey:@"Location"] UTF8String]);
    printf("New Activation Key value: %s\n\n", [[[[[request responseHeaders] objectForKey:@"Location"] componentsSeparatedByString:@"/"] lastObject] UTF8String]);
	
    //Further process the activation key
    int ret = verify_and_invalidate_activation_key([NSURL URLWithString:[[request responseHeaders] objectForKey:@"Location"]]);
    
    [pool drain];
	return ret;
}

static int verify_and_invalidate_activation_key(NSURL *url) {
    printf("Processing the activation key..............\n\n");
    ASIHTTPRequest *request = setup_request(url);
    //Verify - HTTP GET
    [request setRequestMethod:@"GET"];
    [request start];
    
    NSError *error = [request error];
    if(error) {
        NSLog([error description]);
        return 1;
    }
    
    if([request responseStatusCode] == 204) {
        printf("The activation code is valid!\n");
    }
    else {
        printf([[request responseString] UTF8String]);
    }
    
    //Now invalidate - HTTP DELETE
    request = setup_request(url);
    [request setRequestMethod:@"DELETE"];
    [request start];
    
    error = [request error];
    if(error) {
        NSLog([error description]);
        return 1;
    }
    
    printf("The invalidation of the key returned HTTP status code: %d\n", [request responseStatusCode]);
    return 0;
}

static ASIHTTPRequest * setup_request(NSURL *url) {
    ASIHTTPRequest *request = [[[ASIHTTPRequest alloc] initWithURL:url] autorelease];
    [request setUsername:@"rats"];
    [request setPassword:@"RAT1234"];
    [request setShouldRedirect:NO];
    [request setValidatesSecureCertificate:NO];
    return request;
}